README.md
![npm version](https://img.shields.io/npm/v/cds-pg)
![Package Build](https://github.com/sapmentors/cds-pg/workflows/Node.js%20Package/badge.svg)

# `cds-pg` - PostgreSQL adapter for SAP CDS (CAP)

First a big thank you to our contributors:

[![Contributors Display](https://badges.pufler.dev/contributors/sapmentors/cds-pg?size=50&padding=5&bots=false)](https://github.com/sapmentors/cds-pg/graphs/contributors)

This node module provides an adapter to the PostgreSQL database.

For a short introduction on the background of this project you can check out [a short video](https://www.youtube.com/watch?v=b9sPczwYN5Q&t=2310s) that has been captured as part of the SAP devtoberfest.

## Current status

**_cds-pg_ is ready to be used!**
Still, there's some gaps left to fill - note the list below and please see [`CONTRIBUTING.md`](./docs/CONTRIBUTING.md) for how to contribute additional capabilities!

Also checkout the following blog posts on how to get started using `cds-pg` in your local development environment and on SAP Business Technology Platform (BTP), Cloud Foundry:

- [Setup an example project using `cds-pg`](https://blogs.sap.com/2020/11/16/getting-started-with-cap-on-postgresql-node.js/)
- [Prepare and deploy the project to SAP BTP CF](https://blogs.sap.com/2020/11/30/run-and-deploy-cap-with-postgresql-on-sap-cloud-platform-cloud-foundry-node.js/)

### TODO

- [x] implement basic `SELECT|READ`(~ OData `GET`)
- [x] implement basic `INSERT|CREATE`(~ OData `POST`)
- [x] implement basic `UPDATE`(~ OData `PUT|PATCH`)
- [x] implement basic `DELETE`(~ OData `DELETE`)
- [x] map OData to PostgreSQL vocabulary
- [x] implement basic cds deployment
- [x] use default query builders for UPDATE/DELETE
- [ ] add support for multitenancy (see [issue #25](https://github.com/sapmentors/cds-pg/issues/25))
  - [x] support multiple schemas (at config time)
  - [x] support queries with tenant info (at runtime)
- [ ] add support for [full OData vocabulary](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html)
  - [x] system query options (`$filter`, `$expand`, `$select`)
  - [x] string functions
  - [ ] collection functions
  - [ ] date + time functions
  - [ ] geo functions
  - [ ] arithmetic operators + functions
- [x] add draft support (see [issue #30](https://github.com/sapmentors/cds-pg/issues/30))
- [x] add advanced deployment model that supports delta handling/migrations (see [issue #27](https://github.com/sapmentors/cds-pg/issues/27))
- [x] support [`virtual` fields](https://cap.cloud.sap/docs/cds/cdl#virtual-elements) (see [issue 204](https://github.com/sapmentors/cds-pg/issues/204))
- [ ] support [`localized` entities](https://cap.cloud.sap/docs/guides/localized-data) (see [issue 196](https://github.com/sapmentors/cds-pg/issues/196))

## Usage in your CAP project

Add this package to your [SAP Cloud Application Programming Model](https://cap.cloud.sap/docs/) project by running:

```bash
npm install cds-pg
```

Then add this configuration to the `cds` section of your `package.json:

```JSON
  "cds": {
    "requires": {
      "db": {
        "kind": "postgres"
      },
      "postgres": {
        "dialect": "plain", // <- for cds >= 5.1
        "impl": "cds-pg",
        "model": [
          "srv"
        ]
      }
    }
  }
```

For local development you can provide the credentials in the file `default-env.json` in the root folder of your project:

```JSON
{
  "VCAP_SERVICES": {
    "postgres": [
      {
        "name": "postgres",
        "label": "postgres",
        "tags": ["plain", "database"],
        "credentials": {
          "host": "localhost",
          "port": "5432",
          "database": "dbname",
          "user": "postgres",
          "password": "postgres",
          "schema":"public"
        }
      }
    ]
  }
}
```

### CDS deployment

`cds-pg` contains the database adapter to translate the incoming requests to PostgreSQL during runtime, but also includes a _quick and dirty_ command to deploy the current data model to the PostgreSQL database specified in `default-env.json`. Initial data will also be filled from the provided `.csv` files following the approach described in [Providing Initial Data](https://cap.cloud.sap/docs/guides/databases#providing-initial-data). Be aware that the existing tables and views are deleted and then re-created according the CDS model, so this should **not be used in production environments**:

`npx cds-pg deploy srv --to db`

For a more sophisticated approach, please check out `cds-dbm` at https://github.com/mikezaschka/cds-dbm, which offers an advanced deployment model including delta handling of data and models!
Please also read the following blogposts for a detailed description, on how `cds-dbm` can be used in combintation with `cds-pg`:

- [Setup an example project using `cds-pg`](https://blogs.sap.com/2020/11/16/getting-started-with-cap-on-postgresql-node.js/)
- [Prepare and deploy the project to SAP BTP CF](https://blogs.sap.com/2020/11/30/run-and-deploy-cap-with-postgresql-on-sap-cloud-platform-cloud-foundry-node.js/)

## Projects using `cds-pg`

- [pg-beershop](https://github.com/gregorwolf/pg-beershop) – An example project with multiple deployment scenarios (Cloud Foundry, Kyma, Azure, Google Cloud Platform)
- [cap-devtoberfest](https://github.com/mikezaschka/cap-devtoberfest) – A simple project showcasing how to run `cds-pg` and `cds-dbm` locally and on SAP BTB Cloud Foundry


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 0.1.36 (2023-03-16)

### 0.1.35 (2023-02-13)


### Features

* @sap/cds 6.5.0 compatibility ([#383](https://github.com/sapmentors/cds-pg/issues/383)) ([28b2c42](https://github.com/sapmentors/cds-pg/commit/28b2c42e7c390eb120e8f8e47ddb8ad7d78f9402))

### 0.1.34 (2023-01-27)

### 0.1.33 (2022-11-29)

### 0.1.32 (2022-11-09)

### 0.1.31 (2022-11-09)


### Features

* align debug behaviour with cap standard ([#359](https://github.com/sapmentors/cds-pg/issues/359)) ([ffe84a3](https://github.com/sapmentors/cds-pg/commit/ffe84a371ccf1b1144de8dacf3da8dbdd2a2e86c))

### 0.1.30 (2022-08-03)


### Bug Fixes

* delete with grant ([#334](https://github.com/sapmentors/cds-pg/issues/334)) ([d72ec82](https://github.com/sapmentors/cds-pg/commit/d72ec829643665145552d20c37a235324f3462ce)), closes [#262](https://github.com/sapmentors/cds-pg/issues/262)

### 0.1.29 (2022-06-13)


### Bug Fixes

* issue [#315](https://github.com/sapmentors/cds-pg/issues/315) lowercase INSERT results ([#316](https://github.com/sapmentors/cds-pg/issues/316)) ([7ad25a1](https://github.com/sapmentors/cds-pg/commit/7ad25a1865a1b479fa0e63f0ed6dcc40e0f45704))

### 0.1.28 (2022-04-04)


### Features

* support @sap/cds 5.9.1  ([#290](https://github.com/sapmentors/cds-pg/issues/290)) ([2ad4acc](https://github.com/sapmentors/cds-pg/commit/2ad4acc6d64fe8a77ef6ee844fbba0fefe073c36))

### 0.1.27 (2022-04-01)

### 0.1.26 (2022-03-19)


### Bug Fixes

* support deep update for one to many composition ([#281](https://github.com/sapmentors/cds-pg/issues/281)) ([5e9e124](https://github.com/sapmentors/cds-pg/commit/5e9e124c88a2e94e49be891c4b2726cd8d6adcc4))

### 0.1.25 (2022-03-19)

### 0.1.24 (2022-03-19)


### Bug Fixes

* return timestamps, not now() for createdat and modifiedat columns ([#278](https://github.com/sapmentors/cds-pg/issues/278)) ([8a2c05b](https://github.com/sapmentors/cds-pg/commit/8a2c05bef8106838ba3228bbf1040e7241ee12d5))

### 0.1.23 (2022-03-16)

### 0.1.22 (2022-01-25)


### Bug Fixes

* return null when SELECT.one returns empty result instead of empty array ([#266](https://github.com/sapmentors/cds-pg/issues/266)) ([99dc308](https://github.com/sapmentors/cds-pg/commit/99dc308109c0455b5b057023f6d2deeace82b733))

### 0.1.21 (2022-01-24)

### 0.1.20 (2021-12-29)

### 0.1.19 (2021-12-29)

### 0.1.18 (2021-12-29)

### 0.1.17 (2021-12-27)

### 0.1.16 (2021-12-08)


### Bug Fixes

* timestamp conversion issue ([#243](https://github.com/sapmentors/cds-pg/issues/243)) ([ed83ae4](https://github.com/sapmentors/cds-pg/commit/ed83ae47a0f867805e44bff85f3213cff15aaa1d))

### 0.1.15 (2021-11-18)

### 0.1.14 (2021-10-19)


### Features

* support virtual fields (fixes [#204](https://github.com/sapmentors/cds-pg/issues/204)) ([#211](https://github.com/sapmentors/cds-pg/issues/211)) ([9a60e57](https://github.com/sapmentors/cds-pg/commit/9a60e57077092fe39f7f01e2351e5b84193a2f64))

### 0.1.13 (2021-10-06)

### 0.1.12 (2021-10-06)

### 0.1.11 (2021-10-06)

### 0.1.10 (2021-09-21)

### 0.1.8 (2021-08-16)


### Features

* support @sap/cds 5.4 ([1755362](https://github.com/sapmentors/cds-pg/commit/17553620ee5f322b2730742527e544369be49573))

### 0.1.7 (2021-08-02)


### Features

* re-using as release commit for cds 5.3 support ([#180](https://github.com/sapmentors/cds-pg/issues/180)) ([a79c37e](https://github.com/sapmentors/cds-pg/commit/a79c37e9ce010e38662deadf4768f979296ef984))

### 0.1.6 (2021-06-29)

### 0.1.5 (2021-06-16)


### Bug Fixes

* draft support ([#159](https://github.com/sapmentors/cds-pg/issues/159)) ([f60c391](https://github.com/sapmentors/cds-pg/commit/f60c3911597908354cb467c1fd110bad0f9ccc08)), closes [#79](https://github.com/sapmentors/cds-pg/issues/79)

### 0.1.4 (2021-06-14)

### 0.1.3 (2021-06-11)

### 0.1.2 (2021-06-09)


### Features

* fill managed fields with username  ([#154](https://github.com/sapmentors/cds-pg/issues/154)) ([0640e13](https://github.com/sapmentors/cds-pg/commit/0640e136cb25743f0171bb9a957d2fe4f9f897c9)), closes [#153](https://github.com/sapmentors/cds-pg/issues/153)

### 0.1.1 (2021-05-31)

## 0.1.0 (2021-05-31)


### ⚠ BREAKING CHANGES

* no more node <12 support!

* require node 12+ for cds5 ([#141](https://github.com/sapmentors/cds-pg/issues/141)) ([14077be](https://github.com/sapmentors/cds-pg/commit/14077be8a50bfe5b922f1e49cb7a3390dd7ce9d0))

### 0.0.51 (2021-05-16)

### 0.0.50 (2021-05-13)

### 0.0.49 (2021-05-11)

### 0.0.48 (2021-05-09)

### 0.0.47 (2021-05-07)

### 0.0.46 (2021-05-05)

### 0.0.45 (2021-05-03)

### 0.0.44 (2021-03-23)


### Bug Fixes

* bump dependencies and fix failing testrun ([#103](https://github.com/sapmentors/cds-pg/issues/103)) ([f74fdec](https://github.com/sapmentors/cds-pg/commit/f74fdec888ca29c008f755984bb604852b5557fa)), closes [#102](https://github.com/sapmentors/cds-pg/issues/102)

### 0.0.43 (2021-03-22)

### 0.0.42 (2021-03-02)

### 0.0.41 (2021-01-08)


### Bug Fixes

* **core:** Remove Composition from Select ([#86](https://github.com/sapmentors/cds-pg/issues/86)) ([329e44a](https://github.com/sapmentors/cds-pg/commit/329e44a9e516681f2b0f316c2fa504840fe4cf61))

### 0.0.40 (2020-12-22)

### 0.0.39 (2020-12-08)

### 0.0.38 (2020-12-08)


### Features

* **cds:** support for 4.4 ([#77](https://github.com/sapmentors/cds-pg/issues/77)) ([b23cb92](https://github.com/sapmentors/cds-pg/commit/b23cb9231bb5c1496005965e966f4abee31b3eb8))

### 0.0.37 (2020-12-02)

### 0.0.36 (2020-12-02)

### 0.0.35 (2020-12-02)

### 0.0.34 (2020-12-02)

### 0.0.33 (2020-12-02)

### 0.0.32 (2020-12-01)

### 0.0.31 (2020-12-01)

### 0.0.30 (2020-11-23)

### 0.0.29 (2020-11-22)


### Features

*  early version of draft support ([#50](https://github.com/sapmentors/cds-pg/issues/50)) ([9e9af3c](https://github.com/sapmentors/cds-pg/commit/9e9af3ce9575445037900d653c0937774d00587b))


lib/pg/Service.js
const { Pool } = require('pg')
const { rewrite } = require('@sap/cds/libx/_runtime/db/generic')
/*eslint no-undef: "warn"*/
/*eslint no-unused-vars: "warn"*/
const cds = global.cds || require('@sap/cds/lib')

// postgres specific execution
const execute = require('./execute')

/*
 * The service
 */
module.exports = class PostgresDatabase extends cds.DatabaseService {
  constructor(...args) {
    super(...args)

    // Cloud Foundry provides the user in the field username the pg npm module expects user
    if (this.options.credentials && this.options.credentials.username) {
      this.options.credentials.user = this.options.credentials.username
    }
    // Special handling for:
    // SAP Cloud Platform - Cloud Foundry - PostgreSQL Hyperscaler Service
    if (this.options.credentials && this.options.credentials.hostname) {
      this.options.credentials.host = this.options.credentials.hostname
    }
    if (this.options.credentials && this.options.credentials.dbname) {
      this.options.credentials.database = this.options.credentials.dbname
    }
    if (this.options.credentials && this.options.credentials.sslrootcert) {
      if (typeof this.options.credentials.sslRequired === 'undefined') {
        this.options.credentials.sslRequired = true
      }
      this.options.credentials.ssl = {
        rejectUnauthorized: false,
        ca: this.options.credentials.sslrootcert
      }
    }
    this._pool = new Pool(this.options.credentials)

    // SET SCHEMA
    if (this.options.credentials && this.options.credentials.schema) {
      this._pool.query(`SET search_path TO '${this.options.credentials.schema}';`)
    }

    // Mapping of the pg specific executions. Inspired by the SQLite and HANA adapters.
    // We call the default run handlers but pass in our custom execution functions.
    this._execute = execute
    this._insert = this._queries.insert(execute.insert)
    this._read = this._queries.read(execute.read, execute.stream)
    this._update = this._queries.update(execute.update, execute.select)
    this._delete = this._queries.delete(execute.delete)
    this._run = this._queries.run(this._insert, this._read, this._update, this._delete, execute.cqn, execute.sql)
  }

  /**
   * Convert the cds compile -to sql output to a PostgreSQL compatible format
   * @see https://www.postgresql.org/docs/13/datatype.html
   *
   * NVARCHAR -> VARCHAR
   * DOUBLE -> NUMERIC(15, 15)
   * BINARY_BLOB -> CHAR
   * BLOB -> BYTEA
   * NCLOB -> TEXT
   * TIMESTAMP_TEXT -> TIMESTAMPTZ
   * TIME_TEXT -> TIME
   * DATE_TEXT -> DATE
   *
   * @param {String} SQL from cds compile -to sql
   * @returns {String} postgresql sql compatible SQL
   */
  cdssql2pgsql(cdssql) {
    let pgsql = cdssql.replace(/NVARCHAR/g, 'VARCHAR')
    pgsql = pgsql.replace(/DOUBLE/g, 'NUMERIC(30, 15)')
    pgsql = pgsql.replace(/BINARY_BLOB/g, 'CHAR')
    pgsql = pgsql.replace(/BLOB/g, 'BYTEA')
    pgsql = pgsql.replace(/NCLOB/g, 'TEXT')
    pgsql = pgsql.replace(/TIMESTAMP_TEXT/g, 'TIMESTAMPTZ')
    pgsql = pgsql.replace(/TIME_TEXT/g, 'TIME')
    pgsql = pgsql.replace(/DATE_TEXT/g, 'DATE')
    return pgsql
  }

  init() {
    /*
     * before
     */
    // ensures the correct model is present (e.g., tenant extensions)
    this.before('*', this._ensureModel)
    // VIRTUAL, MANAGED, Keys and null checks
    this.before(['CREATE', 'UPDATE'], '*', this._input)
    // "flattens" the query
    // and "redirects" modification statements (CUD) from view to actual table
    this.before(['CREATE', 'READ', 'UPDATE', 'DELETE'], '*', rewrite)
    this.before('READ', '*', this._virtual.bind(this))

    /*
     * on
     */
    this.on('CREATE', '*', this._CREATE)
    this.on('READ', '*', this._READ)
    this.on('UPDATE', '*', this._UPDATE)
    this.on('DELETE', '*', this._DELETE)

    /*
     * after
     */
    // nothing

    /*
     * tx
     */
    this.on(['BEGIN', 'COMMIT', 'ROLLBACK'], function (req) {
      return this.dbc.query(req.event)
    })

    /*
     * "final on"
     */
    this.on('*', function (req) {
      return this._run(this.model, this.dbc, req.query || req.event, req)
    })
  }

  /**
   * Retrieve model definition from ApplicationService
   */
  _ensureModel() {
    for (const service of cds.services) {
      if (service instanceof cds.ApplicationService) {
        this.model = service.model
        break
      }
    }
  }

  /**
   * assign request metadata
   * @param {Object} req currently served express http request, enhanced by cds
   */
  setModel(req) {
    this.models = req.context._model
  }

  /*
   * connection
   */
  async acquire(arg) {
    const schema = (typeof arg === 'string' ? arg : arg.user.schema) || undefined
    // const tenant = (typeof arg === 'string' ? arg : arg.user.tenant) || 'anonymous'
    const dbc = await this._pool.connect()
    // SET SCHEMA
    if (this.options.credentials && this.options.credentials.schema && schema) {
      dbc.query(`SET search_path TO '${schema}';`)
    } else if (this.options.credentials && this.options.credentials.schema) {
      dbc.query(`SET search_path TO '${this.options.credentials.schema}';`)
    }
    return dbc
  }

  /**
   * release the query client back to the pool
   * explicitly passing a truthy value
   * see https://node-postgres.com/api/pool#releasecallback
   */
  async release(dbc) {
    await dbc.release(true)
    return 'dummy'
  }

  // if needed
  async disconnect(tenant = 'anonymous') {
    // potential await custom_disconnect_function(tenant)
    super.disconnect(tenant)
  }

  // REVISIT: Borrowed from SQLite service, but needs cleanup
  async deploy(model, options = {}) {
    let createEntities = cds.compile.to.sql(model)
    if (!createEntities || createEntities.length === 0) return // > nothing to deploy

    // Transform to PostgresSQL
    createEntities = createEntities.map((e) => this.cdssql2pgsql(e))

    const dropViews = []
    const dropTables = []
    for (let each of createEntities) {
      const [, table, entity] = each.match(/^\s*CREATE (?:(TABLE)|VIEW)\s+"?([^\s(]+)"?/im) || []

      // CASCADE added, because without cascade it will through errors for dependant views.
      if (table) dropTables.push(`DROP TABLE IF EXISTS ${entity} CASCADE`)
      else dropViews.push(`DROP VIEW IF EXISTS ${entity} CASCADE`)
    }

    if (options.dry) {
      const log = console.log // eslint-disable-line no-console
      for (let {
        DROP: { view }
      } of dropViews) {
        log('DROP VIEW IF EXISTS ' + view + ' CASCADE;')
      }
      log()
      for (let {
        DROP: { entity }
      } of dropTables) {
        log('DROP TABLE IF EXISTS ' + entity + ' CASCADE;')
      }
      log()
      for (let each of createEntities) log(each + ';\n')
      return
    }

    const tx = this.transaction()
    await tx.run(dropViews)
    await tx.run(dropTables)
    await tx.run(createEntities)
    await tx.commit()

    return true
  }

  // Required for CDS 6.x
  getDbUrl() {
    const credentials = this.options.credentials || this.options || {}
    const dbUrl = `postgres://${credentials.host}//${credentials.database}`

    return dbUrl
  }
}
